## *Jeu de données IRVE*
-------------------
# Analyse des relations entre champs (intégrité des données)
------------------------------
## Contexte
- clarification du rôle des modèles de données dans les jeux de données (cf mise à jour récente des [guides data.gouv](https://guides.etalab.gouv.fr/qualite/documenter-les-donnees/))
- intégration d'une propriété "relationship" dans les schémas de données ([issue TableSchema](https://github.com/frictionlessdata/specs/issues/803) en cours de validation)
- création d'outils de contrôle des relations entre champs des jeux de données tabulaires (cf usage ci-dessous)

## Objectifs
- valider sur un cas réel l'utilisation d'un modèle de données en complément d'un schéma de données
- identifier les apports que pourraient avoir les contrôles de validation des relations entre champs

## Résultats
- la formalisation d'un modèle de données facilite la compréhension des données et des relations entre champs
- l'outil de contrôle permet d'améliorer significativement la qualité des données par l'identification d'incohérences de relations
- l'identification des incohérences permet de trouver des stratégies de réduction des écarts (dans l'exemple ci-dessous, on passe 36% d'écart à 3,8 %)  
- l'analyse des données permet de (re)construire le modèle de données qui minimise les incohérences
- les incohérences détectées sur le jeu de données IRVE restent faibles (inférieures à 4% des point de charge documentés - voir chapitre 4)

## Suite à donner
- Mettre à jour, valider et publier le modèle de données IRVE
- Définir les contrôles supplémentaires à intégrer pour toutes nouvelles données ainsi que pour le jeu complet
- Mettre en oeuvre les outils de contrôle

## Evolutions possibles 
- Ajouter dans les guides d'Etalab un guide pour les modèles de données 
- Intégrer dans les schémas de données la propriété "relationship" en cours de validation,
- Définir un indicateur qui mesure l'écart (existant / attendu) des relations entre champs

## Sommaire
*(liens actifs sur jupyter Notebook ou Nbviewer)*
- [1 - modèle de données](#1---modèle-de-données)
- [2 - Initialisation](#2---Initialisation)
- [3 - Séparation des pdc itinerance et hors itinerance](#3---Séparation-des-pdc-itinerance-et-hors-itinerance)
- [4 - Bilan initial intégrité](#4---Bilan-initial-intégrité)
- [5 - Séparation doublons pdc - date de maj](#5---Séparation-doublons-pdc---date-de-maj)
- [6 - Séparation doublons station - date de maj](#6---Séparation-doublons-station---date-de-maj)
- [7 - Synthèse](#7---Synthèse)
- [8 - Exemples d erreurs résiduelles](#8---Exemples-d-erreurs-résiduelles)
- [Annexe : Initialisation du modèle de données 'mermaid'](#Annexe-:-Initialisation-du-modèle-de-données-'mermaid')

Ce Notebook peut être consulté sur [nbviewer](http://nbviewer.org/github/loco-philippe/Environmental-Sensing/tree/main/python/Validation/irve)

données utilisées : https://www.data.gouv.fr/fr/datasets/fichier-consolide-des-bornes-de-recharge-pour-vehicules-electriques/    
fichier : "*consolidation-etalab-schema-irve-statique-v-2.2.0-20230303.csv*"

-----------------------
# 1 - modèle de données

Le modèle de données proposé ci-dessous est construit sur la base du schéma de données mis à disposition et du contenu du jeu de données.    
Il est à consolider en fonction de l'expertise des concepteurs et réutilisateurs (voir [guide méthodologique](https://github.com/loco-philippe/Environmental-Sensing/blob/main/property_relationship/FR_methodology.ipynb)).

*Notation:*
- *M : Mandatory - documentation obligatoire*
- *PK : Primary Key - identifiant unique de l'entité*
- *Root : champ fictif associé à une ligne du tableau*

In [1]:
from base64 import b64encode
from IPython.display import Image, display
with open('IRVE_modele.txt', 'r', encoding="utf-8") as f:
    modele = f.read()
display(Image(url="https://mermaid.ink/img/" + b64encode(modele.encode("ascii")).decode("ascii")))

In [2]:
def analyse_integrite(data, fields, pr=True):
    '''analyse les relations du DataFrame définies dans le dict fields et retourne un dict avec pour chaque contrôle la
    liste des index ko. Les résultats des contrôles sont également ajoutés sous forme de champs booléens à data'''
    analyse = Analysis(data)
    dic_res = analyse.check_relationship(fields)
    data['ok'] = True
    for name, lis in dic_res.items():
        data[name] = True
        data.loc[lis, name] = False
        data['ok'] = data['ok'] & data[name]
        if pr:
            print('{:<50} {:>5}'.format(name, len(data) - data[name].sum()))
    return dic_res

------
# 2 - Initialisation
## initialisation des données
- lecture du fichier issu de l'api

In [3]:
from datetime import datetime
import json
from observation import Ilist, Analysis
import pandas as pd
log = {'date_irve': '2023-03-03', 'file': 'consolidation-etalab-schema-irve-statique-v-2.2.0-20230303.csv',
      'chemin': 'https://raw.githubusercontent.com/loco-philippe/Environmental-Sensing/main/python/Validation/irve/Analyse/'}
irve = pd.read_csv(log['chemin'] + log['file'], sep=',', low_memory=False)
log['len_irve'] = len(irve)
print('nombre de lignes : ', log['len_irve']) 

nombre de lignes :  51276


## schéma de données
Le schéma de données restreint à la propriété 'relationship' et construit à partir du modèle de données est le suivants :

In [4]:
fields = [
 
 # relation unicité des pdl
 { "name": "index",
   "relationship" : { "parent" : "id_pdc_itinerance", "link" : "coupled" }},   
 # relations inter entités
 { "name": "contact_operateur",
   "relationship" : { "parent" : "id_station_itinerance", "link" : "derived" }},
 { "name": "nom_enseigne",
   "relationship" : { "parent" : "id_station_itinerance", "link" : "derived" }},
 { "name": "coordonneesXY",
   "relationship" : { "parent" : "id_station_itinerance", "link" : "derived" }},
 { "name": "id_station_itinerance",
   "relationship" : { "parent" : "id_pdc_itinerance",     "link" : "derived" }},
 # relations intra entité - station
 { "name": "nom_station",
   "relationship" : { "parent" : "id_station_itinerance", "link" : "derived" }},
 { "name": "implantation_station",
   "relationship" : { "parent" : "id_station_itinerance", "link" : "derived" }},
 #{ "name": "date_maj",
 #  "relationship" : { "parent" : "id_station_itinerance", "link" : "derived" }},
 { "name": "nbre_pdc",
   "relationship" : { "parent" : "id_station_itinerance", "link" : "derived" }},
 { "name": "condition_acces",
   "relationship" : { "parent" : "id_station_itinerance", "link" : "derived" }},
 { "name": "horaires",
   "relationship" : { "parent" : "id_station_itinerance", "link" : "derived" }},
 { "name": "station_deux_roues",
   "relationship" : { "parent" : "id_station_itinerance", "link" : "derived" }},
 # relations intra entité - localisation
 { "name": "adresse_station",
   "relationship" : { "parent" : "coordonneesXY",         "link" : "derived" }} ]

# liste des champs liés à un controle (relations) et obligatoires (mandatory)
relations = ['index', 'contact_operateur', 'nom_enseigne', 'coordonneesXY', 'adresse_station', 'id_station_itinerance', 
             'nom_station', 'implantation_station', 'nbre_pdc', 'condition_acces', 'horaires', 'station_deux_roues', 
             'id_pdc_itinerance', 'date_maj', 'last_modified']
mandatory =  ['contact_operateur', 'nom_enseigne', 'coordonneesXY', 'adresse_station', 'id_station_itinerance', 'nom_station',
           'implantation_station', 'nbre_pdc', 'condition_acces', 'horaires', 'station_deux_roues', 'id_pdc_itinerance', 
           'puissance_nominale', 'prise_type_ef', 'prise_type_2', 'prise_type_combo_ccs', 'prise_type_chademo', 
           'prise_type_autre', 'paiement_acte', 'paiement_autre', 'reservation',  'accessibilite_pmr', 'restriction_gabarit', 
           'date_maj', 'last_modified']

-----------------------
## 3 - Séparation des pdc itinerance et hors itinerance
- un peu moins de 1 % des points de charge sont hors itinerance

In [5]:
data = irve
data[['id_station_itinerance','id_pdc_itinerance']] = data[['id_station_itinerance','id_pdc_itinerance']].astype('string')
#data[['last_modified','date_maj']] = data[['last_modified','date_maj']].astype('datetime64')
data['non_concerne'] = data['id_station_itinerance'].str.contains('oncern') | data['id_pdc_itinerance'].str.contains('oncern')

non_concerne = data[data['non_concerne']].reset_index()['index']
itinerance = data[~data['non_concerne']].reset_index()
itinerance_init = itinerance.loc[:, relations]
log['pdc_hors_itinerance'] = len(non_concerne)
log['pdc_en_itinerance'] = len(itinerance)
print('nombre de pdc hors itinerance : ', log['pdc_hors_itinerance'])
print('nombre de pdc en itinerance   : ', log['pdc_en_itinerance'])

nombre de pdc hors itinerance :  607
nombre de pdc en itinerance   :  50669


In [6]:
# arborescence des champs et nombre de valeurs différentes
il = Ilist.obj(itinerance_init)
print(il.tree())

-1: root-derived (50669)
   0 : index (50669)
      1 : contact_operateur (120)
      2 : nom_enseigne (1132)
      3 : coordonneesXY (16849)
      4 : adresse_station (16094)
      5 : id_station_itinerance (18535)
      6 : nom_station (15058)
      7 : implantation_station (5)
      8 : nbre_pdc (43)
      9 : condition_acces (2)
      10: horaires (453)
      11: station_deux_roues (7)
      12: id_pdc_itinerance (42368)
      13: date_maj (565)
      14: last_modified (444)


-----------------------
## 4 - Bilan initial intégrité
- 36 % des lignes sont erronées ( 18 116 )

In [7]:
# séparation données bonnes (itinerance_ok_1) et données résiduelles (itinerance_1)
res = analyse_integrite(itinerance_init, fields)
itinerance_ok_1 = itinerance_init.loc[itinerance_init.ok, relations].reset_index(drop=True)
itinerance_1 = itinerance_init.loc[~itinerance_init.ok, relations].reset_index(drop=True)
itinerance_init = itinerance_init.loc[:, relations]
log['init_ok'] = len(itinerance_ok_1)
log['init_ko'] = len(itinerance_1)
print("\nnombre d'enregistrements sans erreurs : ", log['init_ok'])
print("nombre d'enregistrements avec au moins une erreur : ", log['init_ko'])
print("taux d'erreur : ", round(log['init_ko'] / log['pdc_en_itinerance'] * 100), ' %')

index - id_pdc_itinerance                          16123
contact_operateur - id_station_itinerance          10719
nom_enseigne - id_station_itinerance                7514
coordonneesXY - id_station_itinerance              11825
id_station_itinerance - id_pdc_itinerance            578
nom_station - id_station_itinerance                 1865
implantation_station - id_station_itinerance        1245
nbre_pdc - id_station_itinerance                    1458
condition_acces - id_station_itinerance               35
horaires - id_station_itinerance                    9869
station_deux_roues - id_station_itinerance         10968
adresse_station - coordonneesXY                     1360

nombre d'enregistrements sans erreurs :  32553
nombre d'enregistrements avec au moins une erreur :  18116
taux d'erreur :  36  %


-----------------------
## 5 - Séparation doublons pdc - date de maj
- la moitié des pdc en erreur sont liées aux doublons de pdc
- la suppression des doublons permet de diviser par 5 le nombre de lignes erronnées ( 2 166 )

In [8]:
# séparation doublons pdc (doublons_pdc) et données résiduelles (itinerance_2)
itinerance_1['doublons_pdc'] = itinerance_1.sort_values(by='date_maj').duplicated('id_pdc_itinerance', keep='last')

doublons_pdc = itinerance_1[itinerance_1['doublons_pdc']].loc[:, relations].reset_index(drop=True)['index']
itinerance_2 = itinerance_1[~itinerance_1['doublons_pdc']].loc[:, relations].reset_index(drop=True)
itinerance_1 = itinerance_1.loc[:, relations]
log['doublons_pdc'] = len(doublons_pdc)
log['sans_doublons_pdc'] = len(itinerance_2)
print('nombre de doublons pdc : ', log['doublons_pdc'])
print('nombre de pdc sans doublon   : ', log['sans_doublons_pdc'])

nombre de doublons pdc :  8301
nombre de pdc sans doublon   :  9815


In [9]:
# séparation données bonnes (itinerance_ok_3) et données résiduelles (itinerance_3)
res = analyse_integrite(itinerance_2, fields)
itinerance_ok_3 = itinerance_2.loc[itinerance_2.ok, relations].reset_index(drop=True)
itinerance_3 = itinerance_2.loc[~itinerance_2.ok, relations].reset_index(drop=True)
itinerance_2 = itinerance_2.loc[:, relations]
log['etape3_ok'] = len(itinerance_ok_3)
log['etape3_ko'] = len(itinerance_3)
print("\nnombre d'enregistrements sans erreurs : ", log['etape3_ok'])
print("nombre d'enregistrements avec au moins une erreur : ", log['etape3_ko'])

index - id_pdc_itinerance                              0
contact_operateur - id_station_itinerance            200
nom_enseigne - id_station_itinerance                 457
coordonneesXY - id_station_itinerance                624
id_station_itinerance - id_pdc_itinerance              0
nom_station - id_station_itinerance                  204
implantation_station - id_station_itinerance         136
nbre_pdc - id_station_itinerance                     513
condition_acces - id_station_itinerance                5
horaires - id_station_itinerance                     194
station_deux_roues - id_station_itinerance           521
adresse_station - coordonneesXY                      587

nombre d'enregistrements sans erreurs :  7649
nombre d'enregistrements avec au moins une erreur :  2166


-----------------------
## 6 - Séparation doublons station - date de maj
- 20% des erreurs résiduelles sont liées au mélange d'anciens et de nouveaux pdc
- la suppression des anciens pdc permet de réduire de 25% le nombre de lignes erronnées ( 1 616 )
- les dernières erreurs correspondent à 468 stations associées à 32 opérateurs et sont liées à des causes multiples

In [10]:
# séparation doublons stations (doublons_stat_maj) et données résiduelles (itinerance_4)
itinerance_3['stat_maj'] = itinerance_3.id_station_itinerance + itinerance_3.date_maj
stat_maj_unique = itinerance_3.sort_values(by='stat_maj').drop_duplicates('id_station_itinerance', keep='last')
itinerance_3['last_stat_maj'] = itinerance_3['stat_maj'].isin(stat_maj_unique['stat_maj'])

doublons_stat_maj = itinerance_3[~itinerance_3['last_stat_maj']].loc[:, relations].reset_index(drop=True)['index']
itinerance_4 = itinerance_3[itinerance_3['last_stat_maj']].loc[:, relations].reset_index(drop=True)
itinerance_3 = itinerance_3.loc[:, relations]
log['doublons_station'] = len(doublons_stat_maj)
log['sans_doublons_station'] = len(itinerance_4)
print('nombre de doublons stations : ', log['doublons_station'])
print('nombre de pdc sans doublon   : ', log['sans_doublons_station'])

nombre de doublons stations :  375
nombre de pdc sans doublon   :  1791


In [11]:
# séparation données bonnes (itinerance_ok_5) et données résiduelles (itinerance_5 / itinerance_5_full)
res = analyse_integrite(itinerance_4, fields)
itinerance_ok_5 = itinerance_4.loc[itinerance_4.ok, relations].reset_index(drop=True)
itinerance_5_full = itinerance_4.loc[~itinerance_4.ok].reset_index(drop=True)
itinerance_5 = itinerance_5_full.loc[:, relations]
#itinerance_4 = itinerance_4.loc[:, relations]
log['etape5_ok'] = len(itinerance_ok_5)
log['etape5_ko'] = len(itinerance_5)
print("\nnombre d'enregistrements sans erreurs : ", log['etape5_ok'])
print("nombre d'enregistrements avec au moins une erreur : ", log['etape5_ko'])

index - id_pdc_itinerance                              0
contact_operateur - id_station_itinerance              0
nom_enseigne - id_station_itinerance                 267
coordonneesXY - id_station_itinerance                216
id_station_itinerance - id_pdc_itinerance              0
nom_station - id_station_itinerance                   40
implantation_station - id_station_itinerance          23
nbre_pdc - id_station_itinerance                     452
condition_acces - id_station_itinerance                0
horaires - id_station_itinerance                      53
station_deux_roues - id_station_itinerance           325
adresse_station - coordonneesXY                      414

nombre d'enregistrements sans erreurs :  175
nombre d'enregistrements avec au moins une erreur :  1616


In [12]:
# structure des données résiduelles et nombre de valeurs
il = Ilist.obj(itinerance_5)
print(il.tree())

-1: root-derived (1616)
   0 : index (1616)
      2 : nom_enseigne (317)
         9 : condition_acces (2)
      3 : coordonneesXY (413)
      4 : adresse_station (468)
         1 : contact_operateur (32)
      5 : id_station_itinerance (468)
         13: date_maj (82)
      6 : nom_station (432)
      7 : implantation_station (4)
      8 : nbre_pdc (21)
      10: horaires (49)
      11: station_deux_roues (5)
      12: id_pdc_itinerance (1616)
      14: last_modified (54)


In [13]:
# nb maxi d'erreurs
itinerance_4['somme'] = 0
for name in res.keys():
    itinerance_4['somme'] += 1 - itinerance_4[name]
erreurs = max(itinerance_4['somme'])
maxi = itinerance_4[itinerance_4.somme >= erreurs]
print("nombre d'enregistrements avec ", erreurs, " erreurs : ", len(maxi))

nombre d'enregistrements avec  4  erreurs :  12


-----------------------
## 7 - Synthèse


In [14]:
# consolidation des données
itinerance['doublons_stat_maj'] = itinerance['index'].isin(doublons_stat_maj)
itinerance['doublons_pdc'] = itinerance['index'].isin(doublons_pdc)
itinerance['lignes_a_corriger'] = itinerance['index'].isin(itinerance_5['index'])
itinerance['doublons_a_supprimer'] = itinerance['doublons_stat_maj'] | itinerance['doublons_pdc']
itinerance['lignes_ko'] = itinerance['doublons_a_supprimer'] | itinerance['lignes_a_corriger']
print('total des lignes à corriger : ', itinerance['lignes_a_corriger'].sum())
itinerance_doublons = itinerance[itinerance['doublons_a_supprimer']].reset_index(drop=True)
print('total des doublons à supprimer : ', len(itinerance_doublons))
itinerance_ok = itinerance[~itinerance['lignes_ko']].reset_index(drop=True)
print('nombre de pdc avec controles ok : ', len(itinerance_ok))

total des lignes à corriger :  1616
total des doublons à supprimer :  8676
nombre de pdc avec controles ok :  40377


In [15]:
#fichier csv des lignes résiduelles à traiter (IRVE_itinerance_residuel)
#fichier csv des données itinerance avec indicateur des données à corriger ou à ignorer (IRVE_itinerance_complet)
#fichier csv des données itinerance valides (IRVE_itinerance_valide)
#fichier csv des doublons (IRVE_itinerance_doublons)
extension = log['date_irve'] +'.csv'
itinerance_5_full.to_csv('IRVE_itinerance_residuel' + extension)
itinerance.to_csv('IRVE_itinerance_complet' + extension)
itinerance_ok.to_csv('IRVE_itinerance_valide' + extension)
itinerance_doublons.to_csv('IRVE_itinerance_doublons' + extension)
log['IRVE_itinerance_residuel' + extension] = len(itinerance_5_full)
log['IRVE_itinerance_complet' + extension] = len(itinerance)
log['IRVE_itinerance_valide' + extension] = len(itinerance_ok)
log['IRVE_itinerance_doublons' + extension] = len(itinerance_doublons)

In [16]:
# vérification de l'intégrité
print('\nbilan intégrité :')
res = analyse_integrite(itinerance_ok.loc[:, relations], fields, pr=False)
log['bilan_erreurs'] = sum([len(controle) for controle in res.values()])
log['date'] = datetime.now().isoformat()
print('\nerreurs : ', log['bilan_erreurs'])


bilan intégrité :

erreurs :  0


In [17]:
with open('logfile.txt', 'a', encoding="utf-8") as f:
    f.write(json.dumps(log) + '\n')
print(log)

{'date_irve': '2023-03-03', 'file': 'consolidation-etalab-schema-irve-statique-v-2.2.0-20230303.csv', 'chemin': 'https://raw.githubusercontent.com/loco-philippe/Environmental-Sensing/main/python/Validation/irve/Analyse/', 'len_irve': 51276, 'pdc_hors_itinerance': 607, 'pdc_en_itinerance': 50669, 'init_ok': 32553, 'init_ko': 18116, 'doublons_pdc': 8301, 'sans_doublons_pdc': 9815, 'etape3_ok': 7649, 'etape3_ko': 2166, 'doublons_station': 375, 'sans_doublons_station': 1791, 'etape5_ok': 175, 'etape5_ko': 1616, 'IRVE_itinerance_residuel2023-03-03.csv': 1616, 'IRVE_itinerance_complet2023-03-03.csv': 50669, 'IRVE_itinerance_valide2023-03-03.csv': 40377, 'IRVE_itinerance_doublons2023-03-03.csv': 8676, 'bilan_erreurs': 0, 'date': '2023-04-17T23:49:47.567173'}


In [18]:
# structure des données bonnes
il = Ilist.obj(itinerance_ok.loc[:, mandatory])
print(il.tree())

-1: root-derived (40377)
   11: id_pdc_itinerance (40377)
      4 : id_station_itinerance (18006)
         1 : nom_enseigne (818)
         2 : coordonneesXY (14075)
            0 : contact_operateur (103)
            3 : adresse_station (13491)
               8 : condition_acces (2)
         5 : nom_station (14149)
         6 : implantation_station (5)
         7 : nbre_pdc (42)
         9 : horaires (411)
         10: station_deux_roues (6)
         19: paiement_autre (7)
      12: puissance_nominale (79)
      13: prise_type_ef (8)
      14: prise_type_2 (8)
      15: prise_type_combo_ccs (8)
      16: prise_type_chademo (7)
      17: prise_type_autre (7)
      18: paiement_acte (8)
      20: reservation (6)
      21: accessibilite_pmr (4)
      22: restriction_gabarit (84)
      23: date_maj (428)
      24: last_modified (361)


-----------------------
## 8 - Exemples d erreurs résiduelles 

### Erreurs multiples 
- exemple avec le nombre maximal d'erreurs (4) (2 stations soit 12 pdc avec : adresse, nom, nbre_pdc et deux-roues erronés)

In [19]:
maxi.loc[:, relations]

index              contact_operateur nom_enseigne         coordonneesXY  \
1536  47980  assistance.technique@tevgo.fr       Belib'   [2.314663, 48.8737]   
1537  47981  assistance.technique@tevgo.fr       Belib'   [2.314663, 48.8737]   
1538  47982  assistance.technique@tevgo.fr       Belib'   [2.314663, 48.8737]   
1539  47983  assistance.technique@tevgo.fr       Belib'   [2.314663, 48.8737]   
1540  47984  assistance.technique@tevgo.fr       Belib'   [2.314663, 48.8737]   
1541  47985  assistance.technique@tevgo.fr       Belib'   [2.314663, 48.8737]   
1648  49154  assistance.technique@tevgo.fr       Belib'  [2.360503, 48.86502]   
1649  49155  assistance.technique@tevgo.fr       Belib'  [2.360503, 48.86502]   
1650  49156  assistance.technique@tevgo.fr       Belib'  [2.360503, 48.86502]   
1651  49157  assistance.technique@tevgo.fr       Belib'  [2.360503, 48.86502]   
1652  49158  assistance.technique@tevgo.fr       Belib'  [2.360503, 48.86502]   
1653  49159  assistance.technique@tevgo.fr       Belib'  [2.360503, 48.86502]   

                    adresse_station id_station_itinerance  \
1536  37 Rue la Boétie 75008 Paris          FRV75PPX0811   
1537  37 Rue la Boétie 75008 Paris          FRV75PPX0811   
1538  37 Rue la Boétie 75008 Paris          FRV75PPX0811   
1539  37 Rue la Boétie 75008 Paris          FRV75PPX0811   
1540  37 Rue la Boétie 75008 Paris          FRV75PPX0811   
1541   37 Rue la Boétie 75008 Paris          FRV75PPX0811   
1648     18 Rue Perrée 75003 Paris          FRV75PPX0304   
1649     18 Rue Perrée 75003 Paris          FRV75PPX0304   
1650     18 Rue Perrée 75003 Paris          FRV75PPX0304   
1651     18 Rue Perrée 75003 Paris          FRV75PPX0304   
1652      18 Rue Perrée 75003 Paris          FRV75PPX0304   
1653     18 Rue Perrée 75003 Paris          FRV75PPX0304   

                    nom_station implantation_station  nbre_pdc  \
1536  Paris | Rue la Boétie 37               Voirie         5   
1537  Paris | Rue la Boétie 37               Voirie         5   
1538  Paris | Rue la Boétie 37               Voirie         5   
1539  Paris | Rue la Boétie 37               Voirie         5   
1540  Paris | Rue la Boétie 37               Voirie         5   
1541   Paris | Rue la Boétie 37               Voirie         6   
1648     Paris | Rue Perrée 18               Voirie         5   
1649     Paris | Rue Perrée 18               Voirie         5   
1650     Paris | Rue Perrée 18               Voirie         5   
1651     Paris | Rue Perrée 18               Voirie         5   
1652      Paris | Rue Perrée 18               Voirie         6   
1653     Paris | Rue Perrée 18               Voirie         5   

     condition_acces horaires station_deux_roues id_pdc_itinerance  \
1536     Accès libre     24/7               True     FRV75EPX08112   
1537     Accès libre     24/7               True     FRV75EPX08111   
1538     Accès libre     24/7              False     FRV75EPX08114   
1539     Accès libre     24/7              False     FRV75EPX08113   
1540     Accès libre     24/7              False     FRV75EPX08115   
1541     Accès libre     24/7              False     FRV75EPX08116   
1648     Accès libre     24/7              False     FRV75EPX03041   
1649     Accès libre     24/7              False     FRV75EPX03042   
1650     Accès libre     24/7              False     FRV75EPX03043   
1651     Accès libre     24/7              False     FRV75EPX03044   
1652     Accès libre     24/7               True     FRV75EPX03046   
1653     Accès libre     24/7               True     FRV75EPX03045   

        date_maj               last_modified  
1536  2022-12-16  2023-03-03T17:35:31.881000  
1537  2022-12-16  2023-03-03T17:35:31.881000  
1538  2022-12-16  2023-03-03T17:35:31.881000  
1539  2022-12-16  2023-03-03T17:35:31.881000  
1540  2022-12-16  2023-03-03T17:35:31.881000  
1541  2022-12-16  2023-03-03T17:35:31.881000  
1648  2022-12-16  2023-03-03T17:35:31.881000  
1649  2022-1

### Cohérence implantation_station - id_station
- 23 pdc sont liés à une erreur de choix d'implantation (3 stations, 1 opérateur)

In [20]:
itinerance_4.loc[~itinerance_4['implantation_station - id_station_itinerance'], relations]

index     contact_operateur nom_enseigne             coordonneesXY  \
1363  43183  contact@qovoltis.com     QOVOLTIS  [-2.7649063, 48.4941535]   
1364  43184  contact@qovoltis.com     QOVOLTIS  [-2.7649063, 48.4941535]   
1365  43185  contact@qovoltis.com     QOVOLTIS  [-2.7649063, 48.4941535]   
1366  43186  contact@qovoltis.com     QOVOLTIS      [-0.69173, 44.83914]   
1367  43187  contact@qovoltis.com     QOVOLTIS      [-0.69173, 44.83914]   
1368  43188  contact@qovoltis.com     QOVOLTIS      [-0.69173, 44.83914]   
1369  43190  contact@qovoltis.com     QOVOLTIS      [-0.69173, 44.83914]   
1372  43282  contact@qovoltis.com     QOVOLTIS      [-0.69173, 44.83914]   
1373  43283  contact@qovoltis.com     QOVOLTIS  [-2.7649063, 48.4941535]   
1375  43343  contact@qovoltis.com     QOVOLTIS   [1.8088319, 43.6986226]   
1379  43579  contact@qovoltis.com     QOVOLTIS      [-0.69173, 44.83914]   
1380  43585  contact@qovoltis.com     QOVOLTIS   [1.8088319, 43.6986226]   
1381  43586  contact@qovoltis.com     QOVOLTIS   [1.8088319, 43.6986226]   
1382  43587  contact@qovoltis.com     QOVOLTIS   [1.8088319, 43.6986226]   
1383  43588  contact@qovoltis.com     QOVOLTIS   [1.8088319, 43.6986226]   
1384  43589  contact@qovoltis.com     QOVOLTIS   [1.8088319, 43.6986226]   
1385  43590  contact@qovoltis.com     QOVOLTIS   [1.8088319, 43.6986226]   
1386  43595  contact@qovoltis.com     QOVOLTIS      [-2.76491, 48.49415]   
1387  43596  contact@qovoltis.com     QOVOLTIS      [-2.76491, 48.49415]   
1393  43611  contact@qovoltis.com     QOVOLTIS      [-2.76491, 48.49415]   
1394  43613  contact@qovoltis.com     QOVOLTIS      [-2.76491, 48.49415]   
1395  43614  contact@qovoltis.com     QOVOLTIS      [-2.76491, 48.49415]   
1397  43638  contact@qovoltis.com     QOVOLTIS      [-2.76491, 48.49415]   

                                        adresse_station id_station_itinerance  \
1363  52 RUE DU GUÉ LAMBERT Hôtel Kyriad Direct Sain...         FRQOVP2200002   
1364  52 RUE DU GUÉ LAMBERT Hôtel Kyriad Direct Sain...         FRQOVP2200002   
1365  52 RUE DU GUÉ LAMBERT Hôtel Kyriad Direct Sain...         FRQOVP2200002   
1366  19 AVENUE NEIL ARMSTRONG, Le Caucase 33700 MER...         FRQOVP3300015   
1367  19 AVENUE NEIL ARMSTRONG, Le Caucase 33700 MER...         FRQOVP3300015   
1368  19 AVENUE NEIL ARMSTRONG, Le Caucase 33700 MER...         FRQOVP3300015   
1369  19 AVENUE NEIL ARMSTRONG, Le Caucase 33700 MER...         FRQOVP3300015   
1372  19 AVENUE NEIL ARMSTRONG, Le Caucase 33700 MER...         FRQOVP3300015   
1373  52 RUE DU GUÉ LAMBERT Hôtel Kyriad Direct Sain...         FRQOVP2200002   
1375             1 AVENUE GEORGES POMPIDOU 81500 LAVAUR         FRQOVP8100001   
1379  19 AVENUE NEIL ARMSTRONG, Le Caucase 33700 MER...         FRQOVP3300015   
1380             1 AVENUE GEORGES POMPIDOU 81500 LAVAUR         FRQOVP8100001   
1381             1 AVENUE GEORGES POMPIDOU 81500 LAVAUR         FRQOVP8100001   
1382             1 AVENUE GEORGES POMPIDOU 81500 LAVAUR         FRQOVP8100001   
1383             1 AVENUE GEORGES POMPIDOU 81500 LAVAUR         FRQOVP8100001   
1384             1 AVENUE GEORGES POMPIDOU 81500 LAVAUR         FRQOVP8100001   
1385             1 AVENUE GEORGES POMPIDOU 81500 LAVAUR         FRQOVP8100001   
1386  52 RUE DU GUÉ LAMBERT Hôtel Kyriad Direct Sain...         FRQOVP2200002   
1387  52 RUE DU GUÉ LAMBERT Hôtel Kyriad Direct Sain...         FRQOVP2200002   
1393  52 RUE DU GUÉ LAMBERT Hôtel Kyriad Direct Sain...         FRQOVP2200002   
1394  52 RUE DU GUÉ LAMBERT Hôtel Kyriad Direct Sain...         FRQOVP2200002   
1395  52 RUE DU GUÉ LAMBERT Hôtel Kyriad Direct Sain...         FRQOVP2200002   
1397  52 RUE DU GUÉ LAMBERT Hôtel Kyriad Direct Sain...         FRQOVP2200002   

                                        nom_station  \
1363  52 RUE DU GUÉ LAMBERT Hôtel Kyriad Direct Sai   
1364  52 RUE DU GUÉ LAMBERT Hôtel Kyriad Direct Sai   
1365  52 RUE DU GUÉ LAMBERT Hôtel Kyriad Direct Sai   
1366  19 AVENUE 

### Cohérence nom_station - id_station
- 40 pdc sont associés à une station avec un nom non cohérent

In [21]:
itinerance_4.loc[~itinerance_4['nom_station - id_station_itinerance'], relations]

index              contact_operateur              nom_enseigne  \
81     1752   service-recharge@soregies.fr               ALTERBASE86   
82     1753   service-recharge@soregies.fr               ALTERBASE86   
107    2862  support@lastmilesolutions.com       Last Mile Solutions   
108    2864  support@lastmilesolutions.com       Last Mile Solutions   
109    2877  support@lastmilesolutions.com       Last Mile Solutions   
110    2878  support@lastmilesolutions.com       Last Mile Solutions   
254    6412         support@alizecharge.fr  CPO Alizé Liberté Public   
256    6431         support@alizecharge.fr  CPO Alizé Liberté Public   
257    6435         support@alizecharge.fr  CPO Alizé Liberté Public   
310    7672         support@alizecharge.fr  CPO Alizé Liberté Public   
311    7693         support@alizecharge.fr  CPO Alizé Liberté Public   
822   22111    pascal.lhermitte@siege27.fr                   SIEGE87   
834   22125    pascal.lhermitte@siege27.fr                   SIEGE88   
861   22152    pascal.lhermitte@siege27.fr                   SIEGE49   
862   22153    pascal.lhermitte@siege27.fr                   SIEGE50   
1362  43174           contact@qovoltis.com                  QOVOLTIS   
1370  43214           contact@qovoltis.com                  QOVOLTIS   
1371  43216           contact@qovoltis.com                  QOVOLTIS   
1374  43288           contact@qovoltis.com                  QOVOLTIS   
1398  43639           contact@qovoltis.com                  QOVOLTIS   
1399  43640           contact@qovoltis.com                  QOVOLTIS   
1400  43643           contact@qovoltis.com                  QOVOLTIS   
1536  47980  assistance.technique@tevgo.fr                    Belib'   
1537  47981  assistance.technique@tevgo.fr                    Belib'   
1538  47982  assistance.technique@tevgo.fr                    Belib'   
1539  47983  assistance.technique@tevgo.fr                    Belib'   
1540  47984  assistance.technique@tevgo.fr                    Belib'   
1541  47985  assistance.technique@tevgo.fr                    Belib'   
1555  48380  assistance.technique@tevgo.fr                    Belib'   
1570  48414  assistance.technique@tevgo.fr                    Belib'   
1571  48415  assistance.technique@tevgo.fr                    Belib'   
1616  48506  assistance.technique@tevgo.fr                    Belib'   
1617  48507  assistance.technique@tevgo.fr                    Belib'   
1618  48509  assistance.technique@tevgo.fr                    Belib'   
1648  49154  assistance.technique@tevgo.fr                    Belib'   
1649  49155  assistance.technique@tevgo.fr                    Belib'   
1650  49156  assistance.technique@tevgo.fr                    Belib'   
1651  49157  assistance.technique@tevgo.fr                    Belib'   
1652  49158  assistance.technique@tevgo.fr                    Belib'   
1653  49159  assistance.technique@tevgo.fr                    Belib'   

                coordonneesXY  \
81        [0.37178, 46.56869]   
82        [0.37178, 46.56869]   
107     [5.585043, 47.149426]   
108     [5.585043, 47.149426]   
109     [5.585043, 47.149426]   
110     [5.585043, 47.149426]   
254     [1.45026, 43.6040701]   
256     [1.45026, 43.6040701]   
257     [1.45026, 43.6040701]   
310   [1.4503666, 43.6039735]   
311   [1.4503666, 43.6039735]   
822     [1.224898, 49.195263]   
834     [1.224898, 49.195263]   
861     [1.221452, 49.197647]   
862     [1.221452, 49.197647]   
1362  [4.0309288, 49.2849379]   
1370  [4.0309288, 49.2849379]   
1371  [4.0309288, 49.2849379]   
1374  [4.0309288, 49.2849379]   
1398    [4.0303445, 49.28518]   
1399    [4.0303445, 49.28518]   
1400    [4.0303445, 49.28518]   
1536      [2.314663, 48.8737]   
1537      [2.314663, 48.8737]   
1538      [2.314663, 48.8737]   
1539      [2.314663, 48.8737]   
1540      [2.314663, 48.8737]   
1541      [2.314663, 48.8737]   
1555   [2.3077698, 48.849297]   
1570   [2.3077698, 48.849297]   
1571   [2.3077698, 48.849297] 

### Cohérence adresse - coordonnées
- 414 pdc ont une adresse non cohérente avec les coordonnées géographiques

In [22]:
itinerance_4.loc[~itinerance_4['adresse_station - coordonneesXY'], relations]

index                        contact_operateur  \
0        10                   info@rossinienergy.com   
1        11                   info@rossinienergy.com   
2        12                   info@rossinienergy.com   
3        13                   info@rossinienergy.com   
4        17                   info@rossinienergy.com   
...     ...                                      ...   
1786  50861  supervision-ev.france@totalenergies.com   
1787  50862  supervision-ev.france@totalenergies.com   
1788  50863  supervision-ev.france@totalenergies.com   
1789  50864  supervision-ev.france@totalenergies.com   
1790  51263                   contact@ubitricity.com   

                   nom_enseigne                          coordonneesXY  \
0                 RossiniEnergy                  [3.262326, 49.856113]   
1                 RossiniEnergy                  [3.262326, 49.856113]   
2                 RossiniEnergy                  [3.262326, 49.856113]   
3                 RossiniEnergy                  [3.262326, 49.856113]   
4                 RossiniEnergy                  [3.262326, 49.856113]   
...                         ...                                    ...   
1786              RIESTER MEAUX  [2.87930851314442, 48.94679007929618]   
1787              RIESTER MEAUX  [2.87930851314442, 48.94679007929618]   
1788              RIESTER MEAUX  [2.87930851314442, 48.94679007929618]   
1789              RIESTER MEAUX  [2.87930851314442, 48.94679007929618]   
1790  Ubitricity France Network                  [0.193942, 49.544211]   

                              adresse_station id_station_itinerance  \
0     4 Avenue Archimède, 02100 Saint-Quentin             FRROSE352   
1     7 Avenue Archimède, 02100 Saint-Quentin             FRROSE352   
2     6 Avenue Archimède, 02100 Saint-Quentin             FRROSE352   
3     5 Avenue Archimède, 02100 Saint-Quentin             FRROSE352   
4     9 Avenue Archimède, 02100 Saint-Quentin             FRROSE351   
...                                       ...                   ...   
1786    52 Rue François De Tessan 77100 Meaux           FRTCBP00702   
1787    52 Rue François De Tessan 77100 Meaux           FRTCBP00702   
1788    51 Rue François De Tessan 77100 Meaux           FRTCBP00702   
1789    50 Rue François De Tessan 77100 Meaux           FRTCBP00702   
1790                        Place Abb� Pierre        FRUBIE10058428   

                       nom_station          implantation_station  nbre_pdc  \
0                     Aventim_PC_1  Parking privé à usage public        10   
1                     Aventim_PC_1  Parking privé à usage public        10   
2                     Aventim_PC_1  Parking privé à usage public        10   
3                     Aventim_PC_1  Parking privé à usage public        10   
4                     Aventim_PC_2  Parking privé à usage public        10   
...                            ...                           ...       ...   
1786                 RIESTER MEAUX  Parking privé à usage public         4   
1787                 RIESTER MEAUX  Parking privé à usage public         4   
1788                 RIESTER MEAUX  Parking privé à usage public         4   
1789                 RIESTER MEAUX  Parking privé à usage public         4   
1790  Centre ville - Montivilliers                Parking public         1   

     condition_acces                                           horaires  \
0        Accès libre                                               24/7   
1        Accès libre                                               24/7   
2        Accès libre                                               24/7   
3        Accès libre                                               24/7   
4        Accès libre                                               24/7   
...              ...                                                ...   
1786     Accès libre  Mo-Fr 08:30-12:00,Mo-Fr 13:30-19:00,Sa 09:00-1...   
1787     Accès libre  Mo-Fr 08:30-12:00,Mo-Fr 13:30-19

- exemple : une station avec plusieurs adresses -> erreur de saisie ?

In [23]:
itinerance_5.loc[itinerance_5.coordonneesXY	 == '[0.193942, 49.544211]', relations]

index       contact_operateur               nom_enseigne  \
83     1866  contact@ubitricity.com  Ubitricity France Network   
1615  51263  contact@ubitricity.com  Ubitricity France Network   

              coordonneesXY    adresse_station id_station_itinerance  \
83    [0.193942, 49.544211]  Place Abbé Pierre        FRUBIE10056568   
1615  [0.193942, 49.544211]  Place Abb� Pierre        FRUBIE10058428   

                       nom_station implantation_station  nbre_pdc  \
83    Centre ville - Montivilliers       Parking public         1   
1615  Centre ville - Montivilliers       Parking public         1   

     condition_acces horaires station_deux_roues id_pdc_itinerance  \
83       Accès libre     24/7              FALSE    FRUBIE10056568   
1615     Accès libre     24/7              FALSE    FRUBIE10058428   

        date_maj               last_modified  
83    2023-01-18  2023-03-03T17:40:10.619000  
1615  2023-01-18  2023-03-03T16:22:10.908000

- exemple : une station avec plusieurs adresses -> erreur de saisie ?

In [24]:
itinerance_5.loc[itinerance_5.coordonneesXY	 == '[2.87930851314442, 48.94679007929618]', relations]

index                        contact_operateur   nom_enseigne  \
1611  50861  supervision-ev.france@totalenergies.com  RIESTER MEAUX   
1612  50862  supervision-ev.france@totalenergies.com  RIESTER MEAUX   
1613  50863  supervision-ev.france@totalenergies.com  RIESTER MEAUX   
1614  50864  supervision-ev.france@totalenergies.com  RIESTER MEAUX   

                              coordonneesXY  \
1611  [2.87930851314442, 48.94679007929618]   
1612  [2.87930851314442, 48.94679007929618]   
1613  [2.87930851314442, 48.94679007929618]   
1614  [2.87930851314442, 48.94679007929618]   

                            adresse_station id_station_itinerance  \
1611  52 Rue François De Tessan 77100 Meaux           FRTCBP00702   
1612  52 Rue François De Tessan 77100 Meaux           FRTCBP00702   
1613  51 Rue François De Tessan 77100 Meaux           FRTCBP00702   
1614  50 Rue François De Tessan 77100 Meaux           FRTCBP00702   

        nom_station          implantation_station  nbre_pdc condition_acces  \
1611  RIESTER MEAUX  Parking privé à usage public         4     Accès libre   
1612  RIESTER MEAUX  Parking privé à usage public         4     Accès libre   
1613  RIESTER MEAUX  Parking privé à usage public         4     Accès libre   
1614  RIESTER MEAUX  Parking privé à usage public         4     Accès libre   

                                               horaires station_deux_roues  \
1611  Mo-Fr 08:30-12:00,Mo-Fr 13:30-19:00,Sa 09:00-1...              false   
1612  Mo-Fr 08:30-12:00,Mo-Fr 13:30-19:00,Sa 09:00-1...              false   
1613  Mo-Fr 08:30-12:00,Mo-Fr 13:30-19:00,Sa 09:00-1...              false   
1614  Mo-Fr 08:30-12:00,Mo-Fr 13:30-19:00,Sa 09:00-1...              false   

     id_pdc_itinerance    date_maj               last_modified  
1611      FRTCBE003959  2023-01-27  2023-03-03T17:35:31.203000  
1612      FRTCBE003958  2023-01-27  2023-03-03T17:35:31.203000  
1613      FRTCBE003955  2023-01-27  2023-03-03T17:35:31.203000  
1614      FRTCBE003954  2023-01-27  2023-03-03T17:35:31.203000

### Cohérence station - coordonnées
- 216 pdc sont associés à des stations avec plusieurs coordonnées

In [25]:
itinerance_4.loc[~itinerance_4['coordonneesXY - id_station_itinerance'], relations][200:215]

index              contact_operateur nom_enseigne  \
1408  44423              contact@evzen.com       Simone   
1409  44424              contact@evzen.com       Simone   
1458  47731  assistance.technique@tevgo.fr       Belib'   
1461  47735  assistance.technique@tevgo.fr       Belib'   
1473  47763  assistance.technique@tevgo.fr       Belib'   
1474  47764  assistance.technique@tevgo.fr       Belib'   
1555  48380  assistance.technique@tevgo.fr       Belib'   
1570  48414  assistance.technique@tevgo.fr       Belib'   
1571  48415  assistance.technique@tevgo.fr       Belib'   
1572  48427  assistance.technique@tevgo.fr       Belib'   
1573  48429  assistance.technique@tevgo.fr       Belib'   
1574  48430  assistance.technique@tevgo.fr       Belib'   
1575  48431  assistance.technique@tevgo.fr       Belib'   
1616  48506  assistance.technique@tevgo.fr       Belib'   
1617  48507  assistance.technique@tevgo.fr       Belib'   

                coordonneesXY  \
1408     [4.427083,43.457513]   
1409     [4.427083,43.457513]   
1458   [2.3664377, 48.865765]   
1461  [2.3664376, 48.8657658]   
1473  [2.3664376, 48.8657658]   
1474   [2.3664377, 48.865765]   
1555   [2.3077698, 48.849297]   
1570   [2.3077698, 48.849297]   
1571   [2.3077698, 48.849297]   
1572   [2.3036203, 48.869644]   
1573   [2.3036203, 48.869644]   
1574   [2.3036203, 48.869644]   
1575  [2.3036203, 48.8696448]   
1616   [2.3077698, 48.849297]   
1617  [2.3077697, 48.8492973]   

                                        adresse_station id_station_itinerance  \
1408  Rue Des Artisans - Borne 2 13460 SAINTES-MARIE...      FRS13PHT2TOP3OOS   
1409  Rue Des Artisans - Borne 2 13460 SAINTES-MARIE...      FRS13PHT2TOP3OOS   
1458                   8 Boulevard Voltaire 75011 Paris          FRV75PPX1105   
1461                   8 Boulevard Voltaire 75011 Paris          FRV75PPX1105   
1473                   8 Boulevard Voltaire 75011 Paris          FRV75PPX1105   
1474                   8 Boulevard Voltaire 75011 Paris          FRV75PPX1105   
1555                   330 Avenue de Ségur 75007 Paris          FRV75PPX0704   
1570                   330 Avenue de Ségur 75007 Paris          FRV75PPX0704   
1571                   330 Avenue de Ségur 75007 Paris          FRV75PPX0704   
1572                 248 Rue Pierre Charron 75008 Paris          FRV75PPX0801   
1573                 248 Rue Pierre Charron 75008 Paris          FRV75PPX0801   
1574                 248 Rue Pierre Charron 75008 Paris          FRV75PPX0801   
1575                 248 Rue Pierre Charron 75008 Paris          FRV75PPX0801   
1616                   330 Avenue de Ségur 75007 Paris          FRV75PPX0704   
1617                    330 Avenue de Ségur 75007 Paris          FRV75PPX0704   

                                  nom_station implantation_station  nbre_pdc  \
1408  Stes-Maries-De-La-Mer, Rue Des Artisans               Voirie         4   
1409  Stes-Maries-De-La-Mer, Rue Des Artisans               Voirie         4   
1458            Paris | Boulevard Voltaire  8               Voirie         2   
1461            Paris | Boulevard Voltaire  8               Voirie         4   
1473            Paris | Boulevard Voltaire  8               Voirie         4   
1474            Paris | Boulevard Voltaire  8               Voirie         2   
1555             Paris | Avenue de Ségur  55               Voirie         5   
1570             Paris | Avenue de Ségur  55               Voirie         5   
1571             Paris | Avenue de Ségur  55               Voirie         5   
1572           Paris | Rue Pierre Charron  62               Voirie         3   
1573           Paris | Rue Pierre Charron  62               Voirie         3   
1574           Paris | Rue Pierre Charron  62               Voirie         3   
1575           Paris | Rue Pierre Charron  62               Voirie         4   
1616             Paris | Avenue de Ségur  55               Voirie         5   
1617              Paris | Avenue de

- exemple : une station avec plusieurs coordonnées -> incompréhension distinction station / pdc

In [26]:
itinerance_5.loc[itinerance_5.id_station_itinerance == 'FR073PCAMAIEUFR', relations]

index         contact_operateur    nom_enseigne         coordonneesXY  \
1169  42830  acelec@acelec-france.com  Camaïeu France   [3.207246,50.68494]   
1170  42831  acelec@acelec-france.com  Camaïeu France  [3.207306,50.684918]   
1171  42832  acelec@acelec-france.com  Camaïeu France  [3.207124,50.684984]   
1172  42833  acelec@acelec-france.com  Camaïeu France  [3.207124,50.684984]   
1173  42834  acelec@acelec-france.com  Camaïeu France  [3.207047,50.685013]   
1174  42835  acelec@acelec-france.com  Camaïeu France  [3.207047,50.685013]   
1175  42836  acelec@acelec-france.com  Camaïeu France  [3.206962,50.685049]   
1176  42837  acelec@acelec-france.com  Camaïeu France  [3.206962,50.685049]   
1179  42840  acelec@acelec-france.com  Camaïeu France   [3.207246,50.68494]   
1180  42841  acelec@acelec-france.com  Camaïeu France  [3.207187,50.684964]   
1181  42844  acelec@acelec-france.com  Camaïeu France  [3.207433,50.684876]   
1182  42845  acelec@acelec-france.com  Camaïeu France  [3.207433,50.684876]   
1183  42846  acelec@acelec-france.com  Camaïeu France  [3.207187,50.684964]   
1184  42847  acelec@acelec-france.com  Camaïeu France   [3.20737,50.684898]   
1185  42848  acelec@acelec-france.com  Camaïeu France   [3.20737,50.684898]   
1186  42849  acelec@acelec-france.com  Camaïeu France  [3.207306,50.684918]   

                         adresse_station id_station_itinerance  \
1169  211 Av. Jules Brame  59100 Roubaix       FR073PCAMAIEUFR   
1170  211 Av. Jules Brame  59100 Roubaix       FR073PCAMAIEUFR   
1171  211 Av. Jules Brame  59100 Roubaix       FR073PCAMAIEUFR   
1172  211 Av. Jules Brame  59100 Roubaix       FR073PCAMAIEUFR   
1173  211 Av. Jules Brame  59100 Roubaix       FR073PCAMAIEUFR   
1174  211 Av. Jules Brame  59100 Roubaix       FR073PCAMAIEUFR   
1175  211 Av. Jules Brame  59100 Roubaix       FR073PCAMAIEUFR   
1176  211 Av. Jules Brame  59100 Roubaix       FR073PCAMAIEUFR   
1179  211 Av. Jules Brame  59100 Roubaix       FR073PCAMAIEUFR   
1180  211 Av. Jules Brame  59100 Roubaix       FR073PCAMAIEUFR   
1181  211 Av. Jules Brame  59100 Roubaix       FR073PCAMAIEUFR   
1182  211 Av. Jules Brame  59100 Roubaix       FR073PCAMAIEUFR   
1183  211 Av. Jules Brame  59100 Roubaix       FR073PCAMAIEUFR   
1184  211 Av. Jules Brame  59100 Roubaix       FR073PCAMAIEUFR   
1185  211 Av. Jules Brame  59100 Roubaix       FR073PCAMAIEUFR   
1186  211 Av. Jules Brame  59100 Roubaix       FR073PCAMAIEUFR   

         nom_station          implantation_station  nbre_pdc condition_acces  \
1169  CAMAÏEU FRANCE  Parking privé à usage public         1     Accès libre   
1170  CAMAÏEU FRANCE  Parking privé à usage public         1     Accès libre   
1171  CAMAÏEU FRANCE  Parking privé à usage public         1     Accès libre   
1172  CAMAÏEU FRANCE  Parking privé à usage public         1     Accès libre   
1173  CAMAÏEU FRANCE  Parking privé à usage public         1     Accès libre   
1174  CAMAÏEU FRANCE  Parking privé à usage public         1     Accès libre   
1175  CAMAÏEU FRANCE  Parking privé à usage public         1     Accès libre   
1176  CAMAÏEU FRANCE  Parking privé à usage public         1     Accès libre   
1179  CAMAÏEU FRANCE  Parking privé à usage public         1     Accès libre   
1180  CAMAÏEU FRANCE  Parking privé à usage public         1     Accès libre   
1181  CAMAÏEU FRANCE  Parking privé à usage public         1     Accès libre   
1182  CAMAÏEU FRANCE  Parking privé à usage public         1     Accès libre   
1183  CAMAÏEU FRANCE  Parking privé à usage public         1     Accès libre   
1184  CAMAÏEU FRANCE  Parking privé à usage public         1     Accès libre   
1185  CAMAÏEU FRANCE  Parking privé à usage public         1     Accès libre   
1186  CAMAÏEU FRANCE  Parking privé à usage public         1     Accès libre   

               horaires station_deux_roues id_pdc_itinerance    date_maj  \
1169  Mo-Su 08:00-08:00              false    FR073E8OV45243  2022-07-21   
1170  Mo-Su 08:00-08:00   

- exemple : une station avec plusieurs coordonnées -> erreur de saisie

In [27]:
itinerance_5.loc[itinerance_5.id_station_itinerance == 'FRV75PPX0704', relations]

index              contact_operateur nom_enseigne  \
1384  48380  assistance.technique@tevgo.fr       Belib'   
1399  48414  assistance.technique@tevgo.fr       Belib'   
1400  48415  assistance.technique@tevgo.fr       Belib'   
1445  48506  assistance.technique@tevgo.fr       Belib'   
1446  48507  assistance.technique@tevgo.fr       Belib'   
1447  48509  assistance.technique@tevgo.fr       Belib'   

                coordonneesXY                   adresse_station  \
1384   [2.3077698, 48.849297]  330 Avenue de Ségur 75007 Paris   
1399   [2.3077698, 48.849297]  330 Avenue de Ségur 75007 Paris   
1400   [2.3077698, 48.849297]  330 Avenue de Ségur 75007 Paris   
1445   [2.3077698, 48.849297]  330 Avenue de Ségur 75007 Paris   
1446  [2.3077697, 48.8492973]   330 Avenue de Ségur 75007 Paris   
1447   [2.3077698, 48.849297]  330 Avenue de Ségur 75007 Paris   

     id_station_itinerance                   nom_station implantation_station  \
1384          FRV75PPX0704  Paris | Avenue de Ségur  55               Voirie   
1399          FRV75PPX0704  Paris | Avenue de Ségur  55               Voirie   
1400          FRV75PPX0704  Paris | Avenue de Ségur  55               Voirie   
1445          FRV75PPX0704  Paris | Avenue de Ségur  55               Voirie   
1446          FRV75PPX0704   Paris | Avenue de Ségur  55               Voirie   
1447          FRV75PPX0704  Paris | Avenue de Ségur  55               Voirie   

      nbre_pdc condition_acces horaires station_deux_roues id_pdc_itinerance  \
1384         5     Accès libre     24/7              False     FRV75EPX07042   
1399         5     Accès libre     24/7              False     FRV75EPX07045   
1400         5     Accès libre     24/7              False     FRV75EPX07046   
1445         5     Accès libre     24/7              False     FRV75EPX07044   
1446         6     Accès libre     24/7              False     FRV75EPX07041   
1447         5     Accès libre     24/7              False     FRV75EPX07043   

        date_maj               last_modified  
1384  2022-12-16  2023-03-03T17:35:31.881000  
1399  2022-12-16  2023-03-03T17:35:31.881000  
1400  2022-12-16  2023-03-03T17:35:31.881000  
1445  2022-12-16  2023-03-03T17:35:31.881000  
1446  2022-12-16  2023-03-03T17:35:31.881000  
1447  2022-12-16  2023-03-03T17:35:31.881000

- exemple : deux stations avec une localisation identique mais deux adresses différentes (pb UTF8) + id identique entre station et pdc

In [28]:
itinerance_5.loc[itinerance_5.coordonneesXY	 == '[0.193942, 49.544211]', relations]

index       contact_operateur               nom_enseigne  \
83     1866  contact@ubitricity.com  Ubitricity France Network   
1615  51263  contact@ubitricity.com  Ubitricity France Network   

              coordonneesXY    adresse_station id_station_itinerance  \
83    [0.193942, 49.544211]  Place Abbé Pierre        FRUBIE10056568   
1615  [0.193942, 49.544211]  Place Abb� Pierre        FRUBIE10058428   

                       nom_station implantation_station  nbre_pdc  \
83    Centre ville - Montivilliers       Parking public         1   
1615  Centre ville - Montivilliers       Parking public         1   

     condition_acces horaires station_deux_roues id_pdc_itinerance  \
83       Accès libre     24/7              FALSE    FRUBIE10056568   
1615     Accès libre     24/7              FALSE    FRUBIE10058428   

        date_maj               last_modified  
83    2023-01-18  2023-03-03T17:40:10.619000  
1615  2023-01-18  2023-03-03T16:22:10.908000

### Incohérence 'nbre_pdc'
- 25% des pdc ont un champ 'nbre_pdc' mal documenté

- exemple : station avec un pdc 'de regroupement' qui évite de documenter plusieurs lignes

In [29]:
itinerance_init.loc[itinerance_init.id_station_itinerance == 'FRLUMEACACIAS11', 
               ['id_pdc_itinerance', 'id_station_itinerance', 'coordonneesXY', 'nbre_pdc']]

id_pdc_itinerance id_station_itinerance           coordonneesXY  nbre_pdc
0   FRLUMEACACIAS11       FRLUMEACACIAS11  [3.1485441,50.7582304]         4

- exemple : station avec plusieurs pdc mais l'id_station est identique à l'id_pdc

In [32]:
itinerance_init.loc[itinerance_init.coordonneesXY	 == '[2.451322, 45.66523]', 
                   ['id_pdc_itinerance', 'id_station_itinerance', 'coordonneesXY', 'nbre_pdc']]

id_pdc_itinerance id_station_itinerance         coordonneesXY  nbre_pdc
50616     FRSDGEMBDG512         FRSDGPMBDG512  [2.451322, 45.66523]         2
50617     FRSDGEMBDG511         FRSDGPMBDG511  [2.451322, 45.66523]         2

- exemple : station avec plusieurs pdc mais le champs a toujours une valeur de 1

In [31]:
itinerance_init.loc[itinerance_init.coordonneesXY	 == '[-1.7548354193520386, 48.125067030488154]', 
                   ['id_pdc_itinerance', 'id_station_itinerance', 'coordonneesXY', 'nbre_pdc']]

id_pdc_itinerance id_station_itinerance  \
1294    FRIENE35353A31          FRIENS35353A   
1295    FRIENE35353A51          FRIENS35353A   
1296    FRIENE35353A41          FRIENS35353A   
1297    FRIENE35353A21          FRIENS35353A   
1298    FRIENE35353A11          FRIENS35353A   

                                  coordonneesXY  nbre_pdc  
1294  [-1.7548354193520386, 48.125067030488154]         1  
1295  [-1.7548354193520386, 48.125067030488154]         1  
1296  [-1.7548354193520386, 48.125067030488154]         1  
1297  [-1.7548354193520386, 48.125067030488154]         1  
1298  [-1.7548354193520386, 48.125067030488154]         1